scripts/run_experiment.sh

```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
+#SBATCH --partition=ephemeral
+#SBATCH --qos=ephemeral
+#SBATCH --ntasks=1
+#SBATCH --mem-per-cpu=8G
+#SBATCH --cpus-per-task=1
+#SBATCH --job-name=ADAM_EXPERIMENT
+#SBATCH --output=data/slurm_logs/%x-%j.out    # %x-%j means JOB_NAME-JOB_ID.
+#SBATCH --requeue
+#SBATCH --exclude="saga01,saga02,saga03,saga04,saga05,saga06,saga07,saga08,saga10,saga11,saga12,saga13,saga14,saga15,saga16,saga17,saga18,saga19,saga20,saga21,saga22,saga23,saga24,saga25,saga26,gaia01,gaia02"
+# The excluded nodes above have slower CPUs on the cluster. This list targets our execution at the CPUs with the highest
+# base & boost frequency
+
+set -euo pipefail
+
+source "$1"/etc/profile.d/conda.sh
+conda activate "$2"
+
+echo "Current node: $(hostname)"
+echo "Current working directory: $(pwd)"
+echo "Starting run at: $(date)"
+echo "Job ID: $SLURM_JOB_ID"
+
+# Run GPT-2 script
+echo
+time PYTHONPATH=. python adam/experiment/log_experiment.py "$3"
+echo
+
+# Finish up the job
+echo "Job finished with exit code $? at: $(date)"
```

Comment:
I think this is subtly not what you would expect, for two reasons:

1. The blank echo clobbers the return code -- yes, even though it's a builtin. 🙃
2. `set -e` means Slurm will fail after `time` if the Python script exits nonzero. So the exit code echo will only ever print "exit code 0". You can fix this by wrapping in an if statement (to override the nonzero exit behavior) and grabbing the exit code ahead of time (see below). Unfortunately, you can't just do `|| true` which does prevent the script crashing but also clobbers the exit code.

(Optionally see [this guide](https://github.com/anordal/shellharden/blob/master/how_to_do_things_safely_in_bash.md) for some more Bash quirks)

```suggestion
if time PYTHONPATH=. python adam/experiment/log_experiment.py "$3"; then
    EXITCODE=0
else
    EXITCODE=$?
fi
echo

# Finish up the job
echo "Job finished with exit code $EXITCODE at: $(date)"
```

Reply:
Huh, this probably means the scripts from the source of this bash script are also wrong.